### DISCUSSION TITLE
JSON editor suggestions

### DISCUSSION THREAD

**Discussion by jumpjack** [ricktu288/ray-optics#153] [Ideas]
I think the JSON editor has great potential, but currently it lacks a very important feature: highlighting the selected object. Currently it highlights the object only upon chnages... when it's too late! ;-) It would be much more useful to see object data before changing it.

Maybe it would be also useful a checkbox or a button to manually start parsing the JSON: I think immediate parsing upon editing is at high risk of crashing everything, in case of typos!

As a sidenote, I think that manually fine-tuning data (both in JSON or in graphic editor) would be much easier if coordinates were approximated to zero decimals.

As a final note, it would be cool if, one day or the other (very low priority...) you could add the ability to associate unique ids (automatically) and custom names (manually) to objects.

**Comment by ricktu288** [comment]
> highlighting the selected object.

The proper approach is actually much more difficult than the current behavior of highlighting the changed part. See https://github.com/ricktu288/ray-optics/discussions/145#discussioncomment-9565642 Currently a hack is to select it and press an arrow key to move the object and immediately move back.

> I think immediate parsing upon editing is at high risk of crashing everything, in case of typos!

I think currently anything that may lead to freezing the app, such as entering an enormously large number, can as well happen in the graphic editor. I think proper approach is to check the range of numbers. Manual parsing feature is also useful, but to add it some additional UI is needed, and it will lead to complicated issues with the undo/redo feature (since the undo chain of JSON and graphic editor will be out of sync). If you want to imitate this feature, a hack is to leave a syntax error (e.g. delete a comma) somewhere and fix that only when you want to parse.

> fine-tuning data (both in JSON or in graphic editor) would be much easier if coordinates were approximated to zero decimals

Approximating numbers in the JSON imply that all the numbers in the underlying scene will be approximated. This may cause issue in some situation, such as simulation that requires exact angles. But for the popup coordinate editor it may be a good idea. Actually the number shown in the object bar are currently approximated.

> it would be cool if, one day or the other (very low priority...) you could add the ability to associate unique ids (automatically) and custom names (manually) to objects.

I have been considering using the `id` approach for the "handle" feature. I think in the far future there will be a GUI sidebar showing list of objects and modules, so that each object can be named. See https://github.com/ricktu288/ray-optics/discussions/141#discussioncomment-9133179.

**Comment by ricktu288** [comment]
The mouse coordinates is now truncated, so new objects directly created with the mouse will have short decimal expansions.

**Comment by ricktu288** [comment]
For the second point, now you can enable Settings -> Show simulator controls and then turn off "Auto Refresh". Then when you edit the JSON, the scene will not be reloaded until you click "Refresh Scene" or "Refresh Simulation".

**Comment by ricktu288** [comment]
Custom name is implemented in the [sidebar branch](https://github.com/ricktu288/ray-optics/tree/sidebar). Will merge to master after the development of the sidebar is complete.

**Reply by ricktu288** [reply]
Merged into master
